Add Collector security documentation (#5209)
open-telemetry/opentelemetry.io: content/en/docs/security/config-best-practices.md

---
title: Collector configuration best practices
linkTitle: Collector config
description: Best practices to securely configure OpenTelemetry Collector.
weight: 20
---

When configuring the OpenTelemetry (OTel) Collector, consider the following
practices to better secure your Collector instance.

## Receivers and exporters

We recommend enabling only the minimum required components for a Collector

> configuration: one receiver and one exporter. Configuring only the minimum set

Reviewer: "one receiver and one exporter" seems very strict, what if the user wants to fork the data to two different targets? (e.g. are we suggesting that forking data is not recommended from security perspective?)

Author: Thanks, all! The content so far has been copied from an earlier draft of this documentation. I think this sentence was an edit of the following statement in the core respository's [README](https://github.com/open-telemetry/opentelemetry-collector/blob/d6f568dd5b6ddfd87a7c6350545cb818f902a06f/docs/security-best-practices.md?plain=1#L63):

> The configuration drives the Collector's behavior and care should be taken to ensure the configuration only enables the minimum set of capabilities and as such exposes the minimum set of required ports.

I will remove the "one receiver and one exporter" restriction from this PR.

Author: I've reworked this whole section (Receivers and exporters) based on the comments. Please let me know what you think. No rush. I still have lots to do. Thanks!

> of components exposes a minimum set of required ports.

Reviewer: "Ports" seem to be too specific, consider rephrasing it to "exposes a minimum attack surface" - it could be TCP/UDP ports, component dependency (SBOM), etc.

Author: Changed in cbf67a9. Thanks!

- Receivers and exporters can be either push- or pull-based. In either case, the
  connection established should be over a secure and authenticated channel.

> - Unused receivers and exporters should be disabled to minimize the attack

Author: Added a recommendation to use `ocb` and clarified that unused components should be removed from configurations. Thanks!

  vector of the Collector.
- Receivers and exporters might expose buffer, queue, payload, and worker
  settings using configuration parameters. If these settings are available, you
  should proceed with caution before modifying the default configuration values.
  Improperly setting these values might expose the OpenTelemetry Collector to

>   additional attack vectors including resource exhaustion.

Reviewer: Resource exhaustion is a great topic, I feel the current description is a bit misleading - even if the settings are correct, OpenTelemetry Collector can still cause resource exhaustion - is this the current state or there is already a solid solution for resource governance?

Author: Hi @reyang, I removed the mention of resource exhaustion here. The topic is addressed in more detail in the Processors section. I'll let you know once I start work on it. Thanks!

Author: Hi again. I updated the resource utilization section in 8605abb.

### Safeguards against denial of service attacks

Bind receivers' servers to addresses that limit connections to authorized users,
so that your Collectors aren't exposed to the public internet or to wider
networks than necessary.

For example, if the OTLP receiver OTLP/gRPC server has only local clients, bind
the `endpoint` setting to `localhost`:

```yaml
receivers:
  otlp:
    protocols:
      grpc:

>         endpoint: localhost:4317

Reviewer: I don't recall the exact details, would this bind to both IPv4 and IPv6, or the Collector has a recommendation regarding how to specify IPv4/v6 explicitly?

Author: I've rewritten the DOS safeguard section as well, with updated information about `localhost`. 

Rather than provide an example of binding to a pod's IP, I linked to the [README section](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/security-best-practices.md#safeguards-against-denial-of-service-attacks). Unless you think that content (Docker, Kubernetes, etc.) is relevant to end users and belongs on the website? I was trying to draw a line between end user and developer content, but I'm happy to make changes.

```

Use `localhost`-like addresses instead of `0.0.0.0`. For more information, see
[CWE-1327: Binding to an Unrestricted IP Address](https://cwe.mitre.org/data/definitions/1327.html).

To change the default endpoint to be `localhost`-bound in all components, enable
the `component.UseLocalHostAsDefaultHost` feature gate. This feature gate will

> be enabled by default in the Collector in a future release.

Reviewer 2: We are in the future already!

### Encryption and authentication

Your OTel Collector configuration should include encryption and authentication.

- For communication encryption, see
  [Configuring certificates](/collector/configuration/#setting-up-certificates).
- For authentication, use the OTel Collector's authentication mechanism, as
  described in [Authentication](/collector/configuration/#authentication).

## Processors

[Processors](/collector/configuration/#processors) sit between receivers and
exporters. They are responsible for processing telemetry before it's analyzed.
From a security perspective, processors are useful in a few ways.

> ### Recommended processors

Reviewer 2: I don't think we should prescribe which processors people should use. We could provide some example configurations for common use-cases though.

Author: Removed the section in favor of examples.

<!--- TODO: SHOULD configure recommended processors. If so, what are they?

redaction processor (never tried myself)
transform processor (great for redacting PIIs and such) -->

### Scrubbing sensitive data

You can use the OpenTelemetry Collector to scrub sensitive data before exporting
it to a backend. Configure the Collector to obfuscate or scrub sensitive data
before exporting.

<!--- TODO: SHOULD configure obfuscation/scrubbing of sensitive metadata. How? Give more details and/or link to an existing document -->

Use OpenTelemetry Collector's `redaction` processor to scrub sensitive data.

<!--- TODO: Give example config for the redaction processor or remove this line. --->

### Safeguarding resource utilization

Processors also offer safeguards for resource utilization. After implementing
safeguards for resource utilization in your
[hosting infrastructure](/security/hosting-best-practices/), make sure your
OpenTelemetry Collector configuration uses these safeguards.

<!-- start same page content in hosting-best-practices -->

> The `batch` and `memory_limiter` processors help ensure the OpenTelemetry

Reviewer 2: Given the discussions we are having around the batch processor, we should probably stop advertising it.

cc @mx-psi

Reviewer 2: After a second thought, perhaps we should stop talking about the memory limiter _processor_ as well, mentioning the _extension_ instead once it gets ready?

https://github.com/open-telemetry/opentelemetry-collector/tree/main/extension/memorylimiterextension

Author: Based on @reyang's earlier comment that resource utilization is an important topic, it seems like we should say something about best practices here. If we don't recommend the batch and memory limiter processors, and the memory limiter extension is still in development, how should we advise users to safeguard resource utilization in their Collector configurations and hosting?

Author: I took a shot at writing this section on my own (8605abb), but it could definitely use a developer's review. 

I mention batching and memory limiting, but I don't call out the processors or extensions specifically. Once the memory limiter extension is ready, we can update with an example.

Reviewer 3: I think a general mention to batching and memory limiting should be more future-proof

Collector is resource efficient and does not run out of memory when overloaded.
These two processors should be enabled on every defined pipeline.

For more information on recommended processors and how to order them in your
configuration, see the
[Collector processor](https://github.com/open-telemetry/opentelemetry-collector/tree/main/processor)
documentation.

<!-- /end same page content in hosting-best-practices -->

## Extensions

While receivers, processors, and exporters handle telemetry directly, extensions
serve different needs.

<!--- TODO: Extensions SHOULD NOT expose sensitive health or telemetry data. How? What can you do? -->

### Health and telemetry

Extensions are available for health check information, Collector metrics and
traces, and generating and collecting profiling data. When enabled with their
default settings, all of these extensions, except the health check extension,
are accessible only locally to the OpenTelemetry Collector. Take care to protect
sensitive information when configuring these extensions for remote access, as
they might expose it accidentally.

### Forwarding

> A forwarding extension is used when you need to collect telemetry that's not

Reviewer 2: I feel like this is misplaced here... I don't think the forwarding extension is related to security, nor is a best practice.

Author: Removed in dbfd1b6. Thanks!

natively supported by the Collector. For example, the `http_forwarder` extension
can receive and forward HTTP payloads. Forwarding extensions are similar to
receivers and exporters so the same security considerations apply.

### Collector's internal telemetry

<!--- INSERT RECOMMENDATIONS HERE. For example:

1. Remove zPages.
1. Remove configuration endpoints.
-->

### Observers

An observer is a component that discovers services in endpoints. Other
components of the OpenTelemetry Collector, such as receivers, can subscribe to
these extensions to be notified of endpoints coming or going.

Observers might require certain permissions in order to discover services. For
example, the `k8s_observer` requires certain RBAC permissions in Kubernetes,
while the `host_observer` requires the OpenTelemetry Collector to run in
privileged mode.

> <!--- But what about Juraci's comment here: https://github.com/open-telemetry/opentelemetry.io/pull/3652/files?diff=unified&w=0#r1417409370 --->

Reviewer 2: It's OK -- what we shouldn't be telling people is to run things as root or blindly disable security protections (like selinux).

### Subprocesses

Extensions can also be used to run subprocesses when the Collector can't

> natively run the collection mechanisms (for example, FluentBit). Subprocesses

Reviewer 2: I'm also not sure where this is coming from: the only component that I know of spawning new processes is the jmx receiver.

Author: This section comes directly from the [security README](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/security-best-practices.md#subprocesses). Should I remove it?

Reviewer 3: I vote for removing this section

Author: Removed in 961ac47.

- TLS certificates including private keys.

You should store sensitive information securely such as on an encrypted
filesystem or secret store. You can use environment variables to handle
sensitive and non-sensitive data as the Collector supports
[environment variable expansion](/docs/collector/configuration/#environment-variables).

### Use encryption and authentication

Your OTel Collector configuration should include encryption and authentication.

- For communication encryption, see
  [Configuring certificates](/docs/collector/configuration/#setting-up-certificates).
- For authentication, use the OTel Collector's authentication mechanism, as
  described in [Authentication](/docs/collector/configuration/#authentication).

### Minimize the number of components

We recommend limiting the set of components in your Collector configuration to
only those you need. Minimizing the number of components you use minimizes the
attack surface exposed.

- Use the
  [OpenTelemetry Collector Builder (`ocb`)](/docs/collector/custom-collector) to
  create a Collector distribution that uses only the components you need.
- If you find that you have unused receivers and exporters, remove them from

>   your configuration.

Author: Sounds good to me.

### Configure with care

Some components can increase the security risk of your Collector pipelines.

> - Receivers and exporters can be push- or pull-based. In either case, you should
>   establish the connection at least over a secure channel, potentially
>   authenticated as well.

Reviewer 3: I don't think the push-pull thing matters much here. There are also other components you may care about properly securing (e.g. if you are using an https confmap provider or an extension):
```suggestion
- Receivers, exporters and other components should establish
  network connections over a secure channel, potentially
  authenticated as well.
```

- Receivers and exporters might expose buffer, queue, payload, and worker
  settings using configuration parameters. If these settings are available, you
  should proceed with caution before modifying the default configuration values.
  Improperly setting these values might expose the OpenTelemetry Collector to
  additional attack vectors.

## Manage specific security risks

Configure your Collector to block these security threats.

### Protect against denial of service attacks

For server-like receivers and extensions, you can protect your Collector from
exposure to the public internet or to wider networks than necessary by binding
these components' endpoints to addresses that limit connections to authorized
users. Try to always use specific interfaces, such as a pod's IP, or `localhost`
instead of `0.0.0.0`. For more information, see
[CWE-1327: Binding to an Unrestricted IP Address](https://cwe.mitre.org/data/definitions/1327.html).

> From Collector v0.110.0, the default endpoints for all servers in Collector
> components are set to `localhost:4317` for `gRPC` ports or `localhost:4318` for
> `http` ports. For earlier versions of the Collector, change the default endpoint
> from `0.0.0.0` to `localhost` in all components by enabling the
> `component.UseLocalHostAsDefaultHost`
> [feature gate](https://github.com/open-telemetry/opentelemetry-collector/tree/main/featuregate).

Reviewer 3: The `4317` and `4318` is specific to the OTLP receiver, not all components.
```suggestion
From Collector v0.110.0, the default host for all servers in Collector components is `localhost`. 
For earlier versions of the Collector, change the default endpoint
from `0.0.0.0` to `localhost` in all components by enabling the
`component.UseLocalHostAsDefaultHost`
[feature gate](https://github.com/open-telemetry/opentelemetry-collector/tree/main/featuregate).
```

accumulating too much data in memory. See the
[`exporterhelper` documentation](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/README.md)
for more information.

Finally, consider using compression with your exporters to reduce the send size
of your data and conserve network and CPU resources. By default, the
[`otlp` exporter](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/otlpexporter)
uses `gzip` compression.

## Extensions

While receivers, processors, and exporters handle telemetry directly, extensions
serve different needs.

<!--- TODO: Extensions SHOULD NOT expose sensitive health or telemetry data. How? What can you do? -->

### Health and telemetry

Extensions are available for health check information, Collector metrics and
traces, and generating and collecting profiling data. When enabled with their
default settings, all of these extensions, except the health check extension,
are accessible only locally to the OpenTelemetry Collector. Take care to protect
sensitive information when configuring these extensions for remote access, as

> they might expose it accidentally.

Reviewer 3: This is mentioned above already in a different fashion, I would just remove this section

Author: Removed in 961ac47.

### Collector's internal telemetry

<!--- INSERT RECOMMENDATIONS HERE. For example:

1. Remove zPages.
1. Remove configuration endpoints.

> -->

Reviewer 3: I would remove this for now and add it back later (we can create an issue for this). I don't think we necessarily need this for the first version

Author: Removed in 961ac47. I also created #5429 so we remember to iterate.
